### PR TITLE
Trustar code review

### PR DESCRIPTION
**Pull request by deepsiIBM**

### CODE REVIEW

File: stix_shifter_modules/trustar/trustar.md [hunk 1]
```diff
@@ -0,0 +1,291 @@
+# Trustar
```
Comment (delliott90) [inline stix_shifter_modules/trustar/trustar.md:1]: This file should be renamed to README.md

**Review by mdazam1942 — CHANGES_REQUESTED** [review]
you need `__init__.py` file inside `stix-shifter/stix_shifter_modules/trustar/stix_translation`

File: stix_shifter_modules/trustar/stix_transmission/api_client.py [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+from stix_shifter_utils.utils.error_response import ErrorResponder
+import json
+import base64
+from stix_shifter_utils.utils import logger
+
+
+class APIClient():
+
+    def __init__(self, connection, configuration):
+        self.base_url = "https://{}".format(connection['host'])
```
Comment (mdazam1942) [inline stix_shifter_modules/trustar/stix_transmission/api_client.py:11]: you don't need to create a base url.

File: stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+  "MALWARE": {
+    "key": "x-trustar-malware.value",
+    "object": "x-trustar-malware"
+  },
+  "URL": {
+    "key": "url.value",
+    "object": "url"
+  },
+  "CVE": {
+    "key": "vulnerability.id",
```
Comment (mdazam1942) [inline stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json:15]: there's no `vulnerability` standard stix object. it needs to go into a custom object

File: stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+  "SHA1": {
+    "key": "file.hashes.SHA1",
+    "object": "file"
+  },
+  "SHA256": {
+    "key": "file.hashes.SHA-256",
+    "object": "file"
+  },
+  "SOFTWARE": [{
+      "key": "file.id",
```
Comment (mdazam1942) [inline stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json:39]: there's no id property in file and software objects. this field can be mapped to software name

File: stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json (hunk 3 above)
Comment (deepsiIBM) [reply stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json]: Done the changes

File: stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json (hunk 4 above)
Comment (deepsiIBM) [reply stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json]: Done the changes

File: stix_shifter_modules/trustar/trustar.md (hunk 1 above)
Comment (deepsiIBM) [reply stix_shifter_modules/trustar/trustar.md]: Done the changes.

File: stix_shifter_modules/trustar/stix_transmission/api_client.py (hunk 2 above)
Comment (deepsiIBM) [reply stix_shifter_modules/trustar/stix_transmission/api_client.py]: Remove the base URL

File: stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json (hunk 4 above)
Comment (mdazam1942) [reply stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json]: software object still has id property that needs to be removed

**Review by mdazam1942 — CHANGES_REQUESTED** [review]
Can you also add stix 2.1 mappings in `trustar/stix_translation/json/stix_2_1`

File: stix_shifter_modules/trustar/stix_transmission/api_client.py [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+        self.headers = dict()
+        self.auth = configuration["auth"]
+        self.client = RestApiClient(connection['host'])
+        self.search_query = ""
```
[the remark this answers is not in the document]
Comment (mdazam1942) [reply stix_shifter_modules/trustar/stix_transmission/api_client.py:14]: still there

File: stix_shifter_modules/trustar/stix_transmission/results_connector.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    def create_results_connection(self, search_id, offset, length):
+        try:
+            min_range = offset
+            max_range = offset + length
+            # Grab the response, extract the response code, and convert it to readable json
+            response_dict = self.api_client.get_search_results(search_id, min_range, max_range)
+            if("code" in response_dict):
+                response_code = response_dict["code"]
+                self.logger.info(response_code)
+                print(response_code)
```
Comment (mdazam1942) [inline stix_shifter_modules/trustar/stix_transmission/results_connector.py:19]: remove print()

File: stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json (hunk 4 above)
Comment (deepsiIBM) [reply stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json]: https://github.com/sathya573/stix-shifter/blob/trustar-CodeReview/stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json

Done the changes.

File: stix_shifter_modules/trustar/stix_transmission/api_client.py [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+        response_data = []
+        resp = self.generate_token
+
+        if resp['code'] == 200:
+            self.token = json.loads(resp['data'])['access_token']
+        else:
+            return response
+
+        self.headers["Authorization"] = "Bearer {}".format(self.token)
+        print(search_id)
```
[the remark this answers is not in the document]
Comment (deepsiIBM) [reply stix_shifter_modules/trustar/stix_transmission/api_client.py]: Done the changes.

File: stix_shifter_modules/trustar/stix_transmission/results_connector.py (hunk 6 above)
Comment (deepsiIBM) [reply stix_shifter_modules/trustar/stix_transmission/results_connector.py]: Done the changes.

**Comment by deepsiIBM** [comment]
> Getting error while translating a combined comparison expression- `[file:name = 'powershell.exe' OR email-addr:value='aa@gmail.com']` Error-
> 
> ```
> {
>     "success": false,
>     "connector": "trustar",
>     "code": "invalid_parameter",
>     "error": "trustar connector error => Error when converting STIX pattern to data source query: Query Not Supported"
> }
> ```

You can try with : [x-trustar-report:title LIKE '%' OR x-trustar-indicator:firstseen LIKE '%'], we are supporting this query.

File: stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json (hunk 4 above)
Comment (mdazam1942) [reply stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json]: still there!
```
{
       "key": "software.id",
       "object": "software",
       "references": "file"
     }
```

**Comment by mdazam1942** [comment]
> > Getting error while translating a combined comparison expression- `[file:name = 'powershell.exe' OR email-addr:value='aa@gmail.com']` Error-
> > ```
> > {
> >     "success": false,
> >     "connector": "trustar",
> >     "code": "invalid_parameter",
> >     "error": "trustar connector error => Error when converting STIX pattern to data source query: Query Not Supported"
> > }
> > ```
> 
> You can try with : [x-trustar-report:title LIKE '%' OR x-trustar-indicator:firstseen LIKE '%'], we are supporting this query.

Okk but the file and email searches suppose to work. they are included in the from_strix mapping

File: stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json (hunk 4 above)
Comment (deepsiIBM) [reply stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json]: update the mapping.
https://github.com/sathya573/stix-shifter/blob/trustar-CodeReview/stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json#L43

**Comment by deepsiIBM** [comment]
> > > Getting error while translating a combined comparison expression- `[file:name = 'powershell.exe' OR email-addr:value='aa@gmail.com']` Error-
> > > ```
> > > {
> > >     "success": false,
> > >     "connector": "trustar",
> > >     "code": "invalid_parameter",
> > >     "error": "trustar connector error => Error when converting STIX pattern to data source query: Query Not Supported"
> > > }
> > > ```
> > 
> > 
> > You can try with : [x-trustar-report:title LIKE '%' OR x-trustar-indicator:firstseen LIKE '%'], we are supporting this query.
> 
> Okk but the file and email searches suppose to work. they are included in the from_strix mapping

Done the changes. Now it is not throwing error.

**Comment by deepsiIBM** [comment]
> Can you also add stix 2.1 mappings in `trustar/stix_translation/json/stix_2_1`

Added  stix 2.1 mappings  for trustar : https://github.com/sathya573/stix-shifter/tree/trustar-CodeReview/stix_shifter_modules/trustar/stix_translation/json/stix_2_1

File: stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    "key": "file.hashes.SHA1",
+    "object": "file"
+  },
+  "SHA256": {
+    "key": "file.hashes.SHA-256",
+    "object": "file"
+  },
+  "SOFTWARE": [{
+      "key": "software.name",
+      "object": "file"
```
Comment (mdazam1942) [inline stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json:40]: this is wrong. object should be something different not "file" that will create a separate software object

File: stix_shifter_modules/trustar/stix_translation/json/stix_2_1/to_stix_map.json [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+    "object": "file"
+  },
+  "SOFTWARE": [{
+      "key": "software.name",
+      "object": "file"
+    },
+    {
+      "key": "software.name",
+      "object": "software",
+      "references": "file"
```
Comment (mdazam1942) [inline stix_shifter_modules/trustar/stix_translation/json/stix_2_1/to_stix_map.json:45]: https://github.com/opencybersecurityalliance/stix-shifter/pull/877#discussion_r953963701

File: stix_shifter_modules/trustar/stix_translation/query_constructor.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+        return "{}".format(raw)
+
+    @staticmethod
+    def _format_equality(value) -> str:
+        return "{}".format(value)
+
+    @staticmethod
+    def _format_like(value) -> str:
+        value = "{value}".format(value=value)
+        return QueryStringPatternTranslator._escape_value(value)
```
Comment (mdazam1942) [inline stix_shifter_modules/trustar/stix_translation/query_constructor.py:58]: are you sure the LIKE operator translation is correct? translating this pattern `[ipv4-addr:value LIKE '127.0.0.%']` returns `"{\"searchTerm\": \"127.0.0. \", \"from\": 1656633600000, \"to\": 1658880300000}"`

the search term seems wrong!

**Comment by mdazam1942** [comment]
Please fix these issues-

1. Combined comparisons expressions do not work correctly for example-
 `[x-trustar-report:title LIKE '%test' OR x-trustar-indicator:firstseen LIKE '%xample']` 
translated into `"{\"title\": \" test\", \"from\": 1661368148099, \"to\": 1661368448099}"`
the second stix object `x-trustar-indicator:firstseen` is disregarded in the translations. 

2. same issue in combined observation expression `[x-trustar-report:title LIKE '%test'] OR [x-trustar-indicator:firstseen LIKE '%xample']`

3. This time ranges throws error `START t'2022-07-01T00:00:00Z' STOP t'2022-07-27T00:05:00Z'`
You need support this timestamp format as well along with `START t'2022-07-01T00:00:00.000Z' STOP t'2022-07-27T00:05:00.000Z'`

**Comment by delliott90** [comment]
@deepsiIBM this PR is pretty old and there are still outstanding comments. Please resolve.

File: stix_shifter_modules/trustar/stix_translation/query_constructor.py (hunk 10 above)
Comment (deepsiIBM) [reply stix_shifter_modules/trustar/stix_translation/query_constructor.py:60]: Done the changes now this query is giving  "{\"searchTerm\": \"127.0.0.%\", \"from\": 1656633600000, \"to\": 1658880300000}"
Screenshot 2022-12-17 at 12 46 34 PM

**Comment by deepsiIBM** [comment]
> Please fix these issues-
> 
> 1. Combined comparisons expressions do not work correctly for example-
>    `[x-trustar-report:title LIKE '%test' OR x-trustar-indicator:firstseen LIKE '%xample']`
>    translated into `"{\"title\": \" test\", \"from\": 1661368148099, \"to\": 1661368448099}"`
>    the second stix object `x-trustar-indicator:firstseen` is disregarded in the translations.
> 2. same issue in combined observation expression `[x-trustar-report:title LIKE '%test'] OR [x-trustar-indicator:firstseen LIKE '%xample']`
> 3. This time ranges throws error `START t'2022-07-01T00:00:00Z' STOP t'2022-07-27T00:05:00Z'`
>    You need support this timestamp format as well along with `START t'2022-07-01T00:00:00.000Z' STOP t'2022-07-27T00:05:00.000Z'`

**Comment by deepsiIBM** [comment]
Resolved the issue : [x-trustar-report:title LIKE '%test' OR x-trustar-indicator:firstseen LIKE '%xample'] , [x-trustar-report:title LIKE '%test'] OR [x-trustar-indicator:firstseen LIKE '%xample'], START t'2022-07-01T00:00:00Z' STOP t'2022-07-27T00:05:00Z' 
Screenshot 2022-12-17 at 12 52 19 PM

**Comment by deepsiIBM** [comment]
>

File: stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+    },
+    {
+      "key": "software.name",
+      "object": "software",
+      "references": "file"
+    }
```
[the remark this answers is not in the document]
Comment (deepsiIBM) [reply stix_shifter_modules/trustar/stix_translation/json/to_stix_map.json:42]: https://github.com/sathya573/stix-shifter/blob/trustar-CodeReview/stix_shifter_modules/trustar/stix_translation/json/stix_2_1/to_stix_map.json#L38

**Comment by delliott90** [comment]
Closing this PR as it's over a year old and seems to have been abandoned. At this point it would be severely out of date with develop.

**Pull request closed without merging**